Cleaner expressions

gwicke opened the pull request:
Implementation of https://phabricator.wikimedia.org/T118700.

gwicke committed:
Support cleaner, un-prefixed path expressions & calls

- Support un-prefixed dotted paths & calls:
  - `call()` expands to `rc.g.call()`
  - `foo[bar]` or `foo.bar` expands to `rm.foo[m.bar]` or `rm.foo.bar`
- Get rid of regexp-based local scope rewrites by leveraging the new
  configuration options in template-expression-compiler v.0.1.5.

gwicke commented:
Confirmed working with https://github.com/wikimedia/restbase/pull/415, which verifies that this code preserves backwards compatibility.

gwicke committed:
Tighten 'simple' path template regexp

The URI constructor doesn't really support templating of partial path
segments, so make sure that we don't use it in that case.

gwicke committed:
Pull in latest template-expression-compiler

gwicke committed:
Add support for {{var}} substitutions

These substitutions don't apply any escaping, and work consistently between
uri & other properties. In the longer term, the goal is to deprecate the use
of {foo} outside of URIs, so that we can eliminate the inconsistent behavior
of that construct between uri & non-uri context.

Task: https://phabricator.wikimedia.org/T118700

gwicke commented:
See also:  wikimedia/restbase#417.

gwicke committed:
Add a `filter` function

We often need to whitelist / filter attributes of an object, so a `filter`
utility for this purpose will be helpful. Currently, this takes an object & an
array of string keys, but we could extend it further in the future to support

- multiple predicate arrays,
- filter predicate functions,
- a key regexp,
- an Array subject.

gwicke committed:
Better error handling & spec fixes

- Bubble up exceptions from template expressions. This avoids silently
  returning 'undefined' requests when undefined nested paths are accessed, for
  example. Access to undefined properties is still supported.
- Bring the implementation of optional path segments closer to
  http://tools.ietf.org/html/rfc6570. Narrow down the use cases supported by
  URI to reflect its lack of proper support for {+foo}.
- Fix several issues in tests: Don't stringify objects, use proper
  assert.deepEqual.
- Distinguish callbacks for the single object return case vs. string
  accumulators.

gwicke committed:
Bump dependencies & release v0.2.8

d00rman commented on lib/reqTemplate.js:
```diff
@@ -69,35 +106,46 @@ function splitAndPrepareTAsseblyTemplate(templateSpec, options) {
+                    inDoubleBrace = true;
+                }
                 startIndex = index + 1;
+
```
Why all these empty lines in the middle?

gwicke replied:
Removed this one in a follow-up. Others below are intentional, to group related functionality a bit.

d00rman commented:
Thinking now about double braces, wouldn't it be better to simply have them in URIs alone? I mena, escaping in other parts of the request makes little to no sense, while on the other hand it requires the user to switch context between request parts, which I think is not that desirable. Compare these two:

``` yaml
reqA:
  uri: {{host}}/{path}
  headers:
    x-first: {request.foo}
  body:
    field: {request.bar}

reqB:
  uri: {{host}}/{path}
  headers:
    x-first: {{request.foo}}
  body:
    field: {{request.bar}}
```

It feels more natural to escape parts of the URI then **not to** escape the headers or body.

gwicke committed:
Better error reporting

In case of a tassembly expansion error, report the original expression source
& tassembly expression, so that users can figure out where the error came
from.

gwicke committed:
Improve error reporting on expansion failure; v0.2.9

gwicke commented:
> Thinking now about double braces, wouldn't it be better to simply have them in URIs alone?

The intention behind the current code is to eventually make single quotes escape everywhere else as well, after a deprecation period & emitting warnings. This way, the behavior will be consistent between uri & other parts, and we can still use the escaped variants to construct URLs, for example in a JSON response containing a href, `Location:` headers, or `ETag`s.

gwicke commented:
Hmmmmmmmm... After thinking about is some more, I'm wondering if we should use a syntax that doesn't require escaping in yaml. The main issue I see is that many characters are reserved there. Main candidates for a prefix would be `$` and `<`. Of these two, `$` (with a space) is somewhat reminiscent of a shell prompt, so perhaps invokes the right intuition?

Example:

``` yaml
request:
  uri: '{{options.host}}/{bar}'
  headers: 
    content-type: $ content-type
    cache-control: $ request.headers.cache-control
  body:
    foo: $ request.body.some.thing[3]
    bar: $ filter(request.body.arr, ['foo', 'bar'])
```

vs.

``` yaml
request:
  uri: '{{options.host}}/{bar}'
  headers: 
    content-type: '{{content-type}}'
    cache-control: '{{request.headers.cache-control}}'
  body:
    foo: '{{request.body.some.thing[3]}}'
    bar: '{{filter(request.body.arr, ['foo', 'bar'])}}'
```

milimetric commented:
Hm.  I think `$ content-type` looks cool, but I also thinks coco looks cool and people almost flayed me for that.  So maybe `$.content-type` is more intuitive.

gwicke commented:
Another option:

``` yaml
request:
  uri: '{{options.host}}/{bar}'
  headers: 
    content-type: <- content-type
    cache-control: <- request.headers.cache-control
  body:
    foo: <- request.body.some.thing[3]
```

berndsi commented:
I agree it does look cool and concise. The `{{ }}` syntax looks more consistent, though, and I feel more confident to know when a variable part is terminated. The `uri` example for the other syntax options seems a bit odd to me in that it introduces the `+`. What does this mean? Why do you need it for `{+options.host}` but not `{bar}`? 
Why do you have `<-` or `$` for most of them but not for the `uri` example?

Ok, take this with a grain of salt since I only looked at this for a minute.

gwicke commented:
@berndsi: Thanks, that's very helpful -- especially the bits that are strange to you.

I also changed the example from under you, sorry. The URI syntax is a subset of http://tools.ietf.org/html/rfc6570, in which `{+foo}` means "literal substitution without escaping", while `{foo}` will percent-escape the parameter. This patch adds support for `{{foo}}` as an alias of `{+foo}`, which is why I changed it back. The question was more about `'{{foo}}'` vs. `$ foo` in attributes that aren't multi-part string / URI templates.

gwicke commented:
Merging in preparation for testing.

berndsi commented:
The `<-` example still shows `{+options.host}` BTW.

gwicke commented:
`$ fixed`
